# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [48]:
# Import your libraries:
import numpy as np
from sklearn.linear_model import LinearRegression
import pandas as pd
from scipy import stats

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function.

In [11]:
# Your code here:
from sklearn.datasets import load_diabetes 
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes. Do this by looking at the `keys()` of this variable.

In [13]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

The next step is to read the description of the dataset. Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable

In [17]:
# Your code here:
diabetes.DESCR

'.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n  :Number of Instances: 442\n\n  :Number of Attributes: First 10 columns are numeric predictive values\n\n  :Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n  :Attribute Information:\n      - age     age in years\n      - sex\n      - bmi     body mass index\n      - bp      average blood pressure\n      - s1      tc, T-Cells (a type of white blood cells)\n      - s2      ldl, low-density lipoproteins\n      - s3      hdl, high-density lipoproteins\n      - s4      tch, thyroid stimulating hormone\n      - s5      ltg, lamotrigine\n      - s6      glu, blood sugar level\n\nNo

What are the variables in this dataset according to the description? List them in the markdown cell below

#### Enter your answer here:




According to this, there are 10 baseline variables. Baseline are all the variables against which results are going to be compared. That is, this 10 variables are quantitatives and categorical variables observed in 442 patients before they start taking the medication. The target, is going to be the disease progression in these 442 patients.

The 10 variables are:
-age

-sex

-body mass index

-avg blood pressure

-6 blood serum measurements

Now explore the data. Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). In the cell below find the shape of the numpy array contained in the data portion of the diabetes variable.

In [23]:
# As we saw before, it has 442 rows (number of patients) and 10 columns (number of variables):
diabetes.data.shape

(442, 10)

# Challenge 2 - Perform Supervised Learning on the Dataset

#### The data has already been split to predictor and response variables. The response variable is in the `target` portion of the variable. 

Given this information, let's apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset. In the cell below, import the linear regression class from sklearn. 

Initialize the model in the variable `diabetes_model`

In [27]:
# Response variable == dependent or outcome variable (y). Predictor variables == independent or input variables (X):
lr = LinearRegression()
X = diabetes.data
y = diabetes.target

diabetes_model = lr.fit(X, y)
predict_coef = diabetes_model.coef_[0]
predict_intercept = diabetes_model.intercept_

In the cell below, fit the model and print the intercept and coefficients of the model. 

In [28]:
# Your code here:
predict_coef, predict_intercept

(-10.012197817470847, 152.1334841628965)

# Bonus Challenge 1 - Conduct a Hypothesis Test on the Model

Once we have generated a linear model, we can test each coefficient using a t-test to see whether the confidence interval for the variable contains zero. We can also perform an overall F test to check whether at least one coefficient is significantly different from zero. 

Refer to the resource in this [link](https://onlinecourses.science.psu.edu/stat501/node/297/) for more details and perform the t-tests for the model above. Additionally, interpret the results and list coefficients are significantly different from zero.


Hint: use the statsmodels package.

Your result should look similar to this:

![ols](../ols-results.png)

In [33]:
# Porque no puedo encontrar el r^2 asi? porque son multiples variables?:

y_predicted = predict_intercept + predict_coef*X
mean_prediction = [y.mean()] * 442
TSS = ((y - mean_prediction) ** 2).sum()
RSS = ((y - y_predicted) ** 2).sum()
r2 = (TSS-RSS)/TSS

In [43]:
import statsmodels.formula.api as smf

In [49]:
stats.ttest_ind(y,X)

Ttest_indResult(statistic=array([41.48784642, 41.48784642, 41.48784642, 41.48784642, 41.48784642,
       41.48784642, 41.48784642, 41.48784642, 41.48784642, 41.48784642]), pvalue=array([1.6921722e-209, 1.6921722e-209, 1.6921722e-209, 1.6921722e-209,
       1.6921722e-209, 1.6921722e-209, 1.6921722e-209, 1.6921722e-209,
       1.6921722e-209, 1.6921722e-209]))

In [44]:
results = smf.ols('y ~ X', data=diabetes).fit()
results.summary()

ERROR! Session/line number was not unique in database. History logging moved to new session 426


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.518
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                     46.27
Date:                Tue, 23 Feb 2021   Prob (F-statistic):           3.83e-62
Time:                        17:39:45   Log-Likelihood:                -2386.0
No. Observations:                 442   AIC:                             4794.
Df Residuals:                     431   BIC:                             4839.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    152.1335      2.576     59.061      0.000     147.071     157.196
X[0]         -10.0122     59.749     -0.168      0.867    -127.448     107.424
X[1]        -239.8191     61.222     -3.917      0.000    -360.151    -119.488
X[2]         519.8398     66.534      7.813      0.000     389.069     650.610
X[3]         324.3904     65.422      4.958      0.000     195.805     452.976
X[4]        -792.1842    416.684     -1.901      0.058   -1611.169      26.801
X[5]         476.7458    339.035      1.406      0.160    -189.621    1143.113
X[6]         101.0446    212.533      0.475      0.635    -316.685     518.774
X[7]         177.0642    161.476      1.097      0.273    -140.313     494.442
X[8]         751.2793    171.902      4.370      0.000     413.409    1089.150
X[9]          67.6254     65.984      1.025      0.306     -62.065     197.316
==============================================================================
Omnibus:                        1.506   Durbin-Watson:                   2.029
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                1.404
Skew:                           0.017   Prob(JB):                        0.496
Kurtosis:                       2.726   Cond. No.                         227.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### El R^2 = 0.518 nos indica que nuestro modelo esta en el limite de ser adecuado. 

# Challenge 2 - Peform Supervised Learning on a Pandas Dataframe

Now that we have looked at data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [108]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Look at the first 5 rows using the `head()` function:

In [109]:
# Your code here:



Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [110]:
type(auto.car_name)

pandas.core.series.Series

In [111]:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [112]:
# Your code here:
res = auto[pd.to_numeric(auto['car_name'], errors='coerce').isnull()]

What is the newest model year and the oldest model year?

In [113]:
auto["model_year"].max(), auto["model_year"].min() 

(82, 70)

Check the dataset for missing values and remove all rows containing at least one missing value.

In [114]:
# Your code here:
auto.dropna(inplace=True)

In [115]:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [116]:
# Your code here: 5 possible values:
auto.cylinders.value_counts().sort_values()

5      3
3      4
6     83
8    103
4    199
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

In [117]:
del auto['car_name']

In [118]:
# Import the necessary function
from sklearn.model_selection import train_test_split

In [119]:
data_y = auto.mpg

In [120]:
del auto['mpg']

In [123]:
data_x = auto

cylinders  displacement  horse_power  weight  acceleration  model_year
0            8         307.0        130.0    3504          12.0          70
1            8         350.0        165.0    3693          11.5          70
2            8         318.0        150.0    3436          11.0          70
3            8         304.0        150.0    3433          12.0          70
4            8         302.0        140.0    3449          10.5          70
..         ...           ...          ...     ...           ...         ...
393          4         140.0         86.0    2790          15.6          82
394          4          97.0         52.0    2130          24.6          82
395          4         135.0         84.0    2295          11.6          82
396          4         120.0         79.0    2625          18.6          82
397          4         119.0         82.0    2720          19.4          82

[392 rows x 6 columns]

In [122]:
data_y

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
393    27.0
394    44.0
395    32.0
396    28.0
397    31.0
Name: mpg, Length: 392, dtype: float64

In [124]:
# Your code here:
X_train, X_test, y_train, y_test = train_test_split(data_x, data_y, test_size=0.2, random_state=None)

In [125]:
len(data_x), len(X_train)

(392, 313)

Now we will the dataset that we processed and peform linear regression on this data to predict the mpg for each vehicle. Initialize the model in the cell below.

In [148]:
# Your code here:
lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [149]:
lr.coef_, lr.intercept_ 

(array([-0.40206423,  0.00928232, -0.00150978, -0.00673149,  0.02041602,
         0.75335251]),
 -13.550533204262159)

# Challenge 3 - Evaluate the Model

the r squared score of a model tells us how much variation is explained by the model. In a typical dataset, most observations differ from the mean. When we create a model, we are trying to generate an equation that will tell us by how much each observation will differ from the mean. Obviously, the vast majority of models are not perfect. They can only predict some of the variation from the mean but not all of it. We attribute the rest of the difference between the actual value and the mean to random error. We would like random error to explain the as little as possible of the variation. This is why the r squared score is an important metric.

In the next cell, compute the r squared score of the model. Do this by first computing the predicted values and assign them to `y_pred`.

In [150]:
y_pred = lr.predict(X_test)

In [151]:
mean_squared_error(y_pred,y_test)

15.777740254920607

In [154]:
r2_score(y_test, y_pred)

0.7784414217318504

#### Our next step is to evaluate the model using the test data. We would like to ensure that our model is not overfitting the data. This means that our model will not be able to generalize well outside of the training data.

In the cell below, use the model to generate the predicted values for the training data and assign them to `y_test_pred`. Compute the r squared score for the test data by comparing the oberserved `y_test` data and the predicted `y_test_pred`.

In [153]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [159]:
# Your code here:
y_test_pred = lr.predict(X_train)

In [156]:
mean_squared_error(y_train, y_test_pred)

11.166337906236196

In [157]:
r2_score(y_train, y_test_pred)

0.8103348775167917

# Challenge 4 - Improve the Model Fit

While the most common way to improve the fit of a model is by using regularization, there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [136]:
X_train, X_test, y_train, y_test = train_test_split(data_x, data_y, test_size=0.1, random_state=None)

In [133]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.svm import SVR
from sklearn.linear_model import SGDRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor

In [140]:
# Your code here:
models = {
    "ridge":Ridge(),
    "lasso":Lasso(),
    "sgd":SGDRegressor(),
    "knn": KNeighborsRegressor(),
    "gradient":GradientBoostingRegressor(),
"svr": SVR()}


Initialize a new model. Name this model `auto_model09`. Fit the model to the new sample data.

In [141]:
# Your code here:
for m in models.values():
    m.fit(X_train,y_train)

Compute the predicted values and r squared score for our new model and new sample data.

In [143]:
# Your code here:
y_pred = {}
for name,m in models.items():
    y_pred[name] = m.predict(X_test)

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [144]:
# Your code here:
mse = {}
for name,y in y_pred.items():
    mse[name] = mean_squared_error(y,y_test)

In [145]:
mse

{'ridge': 15.775999281106646,
 'lasso': 15.83613764466394,
 'sgd': 2.1131680729546078e+30,
 'knn': 19.16685,
 'gradient': 8.237940121459657,
 'svr': 26.33832223218159}

#### So the best fitting model is SGDRegressor

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [200]:
# Your code here:



Fit the model and print the ranking

In [201]:
# Your code here:



Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [202]:
# Your code here:



Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [203]:
# Your code here: 

